### ISSUE TITLE
Allow additional metadata to be passed on upload for addons and hooks

### ISSUE THREAD

**Issue by madduck** [eikek/docspell#2334]
Right now, there is no really good way to attach additional metadata to uploads, e.g. if you know the correspondent of a document. The metadata structure of the upload endpoints have a fixed set of fields, and the upload API calls do not return any information that the caller could use to construct a subsequent API call to update a specific document.

If using an addon for post-processing, or a `job-done` hook, the only somewhat unique information available about the document is the filename, which the addon/hook could use to look up the metadata that was available at upload time, and then do its magic. Since filenames are only unique as absolute paths (and even then, Windows doesn't guarantee that), the filename would have to be set to the absolute path on upload, else there is a risk of collision.

It is possible to change the item name from an addon, or via API call, but it is not possible to change the names of attachments, thus the absolute path would remain permanently exposed/encoded in the DMS.

One possible solution could involve the item ID returned in the upload API response (#2327), but that still means that the caller would have to deal with the asynchronous nature of the processing pipeline.

If the `meta` structure provided during upload could grow another field in which the caller could provide any additional information for addons/hooks to later process, then it would actually be trivial to properly inject available metadata at upload time, and have access to it during post-processing.

**Comment by madduck** [comment]
For the record, I am now abusing the `fileFilter` field of the metadata structure, by appending a Base64-encoded JSON structure like so: `*|SSBhbSBmdWxseSBhd2FyZSB0aGF0IGRlaXRpZXMgYXJlIGtpbGxpbmcga2l0dGVucyBiZWNhdXNlIG9mIG1l`. This is *gross*, I know, but it works for now, as this information can be extracted from addons, and the `job-done` hook, and the likelihood of a conflict i.e. false positive on the file filter is negligible.

**Comment by v6ak** [comment]
What would you do with duplicate uploads? E.g. first upload would have additional metadata `foo` and the second would have additional metadata `bar`. Once it turns out both uploads are the same file, what would be in the metadata?

**Comment by madduck** [comment]
Hey @v6ak, thanks for your interest! If I understand docspell correctly, then a duplicate upload with `skipDuplicates=False` would result in an additional item with its own set of metadata. If `skipDuplicates==True`, then I think the new metadata should just be ignored.

**Comment by madduck** [comment]
A bit more thought about how to get additional information to the item that one uploads, and about the interplay of this issue with #2327:

Right now, you

1. Either react to the processing of a new item with an addon, which then e.g. creates a new row in a database, later to be picked up by another processor,

2. Or react to the processing with a notification web-hook, which does the same.

In both cases, if there is additional information available at upload time, the addon or web-hook need to go and look it up somehow. As you can see from my comment to #2334, the only reliable way to do so right now is by abusing the `fileFilter` field.

If I get the upload ID at upload time (#2327), I can schedule a task to add the additional information later, and add additional information to the task.

If I can pass metadata into the upload so that the web-hook or addon can access it, I don't need the item ID at upload time, because the web-hook or addon has access to the information from the metadata

I think fixing this issue is the way to go. But I still think #2327 would be nice to have anyway. Even though watching job status isn't possible at the moment without polling, knowing the eventual item ID at upload time would allow some sort of closure.

If an upload fails, the client will have an ID that doesn't point to anything. The client can then search for the reason.

**Comment by eikek** [comment]
I think what is ok to do is to be able to thread custom data through the upload and processing that you get back when asynchronously reacting to it. I think, stuffing everything into the upload metadata can lead to a bit surprising behaviour. Meta data may not exist and can be hard to disambiguate automatically given only the name (like correspondents). OTOH perhaps it can be extended to allow metadata that is required to exist or easy to deal with automatically, like custom fields or ids to correspondents.

**Comment by madduck** [comment]
For a start, I think just a single custom field would do fine. It's the user's responsibility to do the right thing with the metadata, including looking it up in the existing data (correspondents), or handling its absence. Small steps…

**Comment by v6ak** [comment]
I might allow to set multiple custom fields, because it doesn't seem harder. Moreover, it seems to be easier from the API design perspective.

**Comment by v6ak** [comment]
I might allow to set multiple custom fields, because it doesn't seem harder. Moreover, it seems to be easier from the API design perspective.

**Comment by madduck** [comment]
We should distinguish between the custom fields that already exist for items in Docspell, and custom fields in the context of this issue. Fundamentally, I don't think there's a difference between adding a single custom field to the existing metadata upload JSON structure, and allowing that structure to have any number of additional fields, except the former nicely namespaces all children. For now, I suggest to keep things simple:

A single field that gets passed to addons and the `job-done` webhook, i.e. wherever the current upload metadata are made available. If the current metadata are stored in the database, then this field will be too. Otherwise I don't really see a reason, to me, the metadata are fleeting and non-ephemeral.

**Comment by madduck** [comment]
I think that #2334 has nothing to do with the database, or with the item, i.e. not the [custom fields that are part of item metadata](https://docspell.org/docs/webapp/customfields/). It is just about being able to add additional information to the [metadata accompanying an upload](https://docspell.org/docs/api/upload/#metadata).

I see why you might want to just bridge to item metadata custom fields there, but then where do you stop? Can/should I be able to set the item date this way? What about correspondents? What about concerning entities? What about due date? What if a custom field doesn't exist for the collective? Should it get created? Should the upload fail, or just ignore the field?

Tags are handled, and that makes it tempting to do other things, sure. However, this request is purely about being able to pass data to the addons/`job-done` hook. If we can do that, it's possible to write and publish parametrised addons to handle setting correspondents, due date, etc.…

**Comment by v6ak** [comment]
OK. I am a bit unsure how to name it. I've asked Bard+ChatGPT and there are some meaningful suggestions:

* upload context
* upload metadata
* upload extras (?)
* use "transient" instead of "upload"

I would probably prefer either "upload context" or "transient metadata". Both names suggest that it is just temporary. None of them explains that one could pass any arbitrary data, but maybe it isn't that important.

**Comment by madduck** [comment]
Well, the structure is already there, and I think it's referred to as "upload metadata". It already defines a number of fields, such as `multiple`, `language` and `tags`. All I am suggesting is that I can also specify `additionalData` or `customData` and pass arbitrary JSON therein.

**Comment by v6ak** [comment]
The upload metadata structure contains some fields that are going to be stored (probably language, tags and folder, maybe more). This is the reason why I find it important to make it clear that this data is just transient and not going to be stored in the DB.

-----

Currently, I have a minor struggle with the type of arbitrary JSON in OpenAPI. I've tried to define it like this:
```
        uploadContext:
          type: object
          additionalProperties: true
          default: null
```

This pass the validation in Swagger editor. However, the codegen task (sbt-openapi-schema, I suppose) fails with this error message:

```
[error] (restapi / openapiCodegen) An object schema with value types `Object` and `AnyRef`, respectively, is not supported.
```

When I try to omit the type specification, I get this error:

```
[error] (webapp / openapiCodegen) Unsupported schema: class Schema {
[error]     type: null
[error]     format: null
…
```
It looks like I am hitting some limitations of the sbt-openapi-schema.

Maybe eikek will know an easy way. ATM, I just use string. While JSON string inside another JSON isn't great, it allows me to move forward and I can change it later.

**Comment by v6ak** [comment]
Well, I've missed ReProcessItem. It tries to reconstruct the upload metadata from DB. So, we can either:

a. lose this data, so that we don't have them when reprocessing
b. save this data in the DB (but in this case, we could use custom fields instead; I agree that there is no perfect solution for situation with non-existent custom field, but maybe log&ignore would be good enough)

**Comment by eikek** [comment]
Hi there, due to lack of time, here some short and quick thoughts 😄 
- I like the idea that we only thread some custom user data through the process
- I would store it into the database as task arguments, just like the other data. The job component is made for that, so there is no spoiling of the "real" item data in the database. And one can argue it belongs to. the job and is required to function properly (when taking addons into account). It should also be quite easy to adopt - no special case necessary.
- I can feel the openapi pain... sorry for that 😅. I'd need to dig in as well. I would also prefer to have an additional property `customData` (or named similar) being arbitrary json. I think there is somewhere an example for that. But I have to search... going with a string for now is fine!

**Comment by v6ak** [comment]
I've pushed a PoC.

* It works for processed items.
* It doesn't work for reprocessed items.
* For all places where I don't know the value of reprocessed items, I add some marker, so that I know why it is empty. This is just for debugging purposes. Before merging, all [markers](https://github.com/v6ak/docspell/commit/63e8a989f798e6115541ce8d9972604b442360d6#diff-34549ea12970cd82a6818ce064fcf9d762ab88d8632906864c6c7634d3b6a5b6R48-R56) will be removed and replaced with `None` (`null` in JSON terminology), or maybe by some other value (in the case of reprocessed items).
* The naming isn't consistent yet.
* Just strings are supported.

**Comment by eikek** [comment]
I think for the "arbitrary json" part in the openapi, the [plugin](https://github.com/eikek/sbt-openapi-schema) needs to be extended.

**Comment by v6ak** [comment]
This should be closed by #2380, shouldn't it?

**Comment by Angelinsky7** [comment]
i have the need to upload a document with custom fields, is this the way of doing it ?

- Passing customData with my upload
- Writing a custom addon to handle those specific customDatas that will set the custom's fields of the uploaded item ?

Thanks to anyone that could help me figure this.
